Describe Matrix URI scheme (#3168)
matrix-org/matrix-doc: content/appendices.md

Room aliases MUST NOT exceed 255 bytes (including the `#` sigil and the
domain).

#### URIs

There are two major kinds of referring to a resource in Matrix: matrix.to
and `matrix:` URI. The specification currently defines both as active/valid

> ways to refer to entities/resources.

Reviewer: No preference/note of deprecation of the former?

Author: The MSC didn't deprecate anything

history (a permalink).

##### Matrix URI scheme

The Matrix URI scheme is defined as follows (`[]` enclose optional parts, `{}`
enclose variables):
```
matrix:[//{authority}/]{type}/{id without sigil}[/{type}/{id without sigil}...][?{query}][#{fragment}]
```

As a schema, this can be represented as:

```
MatrixURI = "matrix:" hier-part [ "?" query ] [ "#" fragment ]
hier-part = [ "//" authority "/" ] path

path = entity-descriptor ["/" entity-descriptor]
entity-descriptor = nonid-segment / type-qualifier id-without-sigil
nonid-segment = segment-nz ; as defined in RFC 3986
type-qualifier = segment-nz "/" ; as defined in RFC 3986
id-without-sigil = string ; as defined in Matrix identifier spec above

query = query-element *( "&" query-item )
query-item = action / routing / custom-query-item
action = "action=" ( "join" / "chat" )
routing = "via=” authority
custom-query-item = custom-item-name "=" custom-item-value
custom-item-name = 1*unreserved ; reverse-DNS name
custom-item-value =
```

Note that this format is deliberately following [RFC 3986](https://tools.ietf.org/html/rfc3986)
to ensure maximum compatibility with existing tooling. This URI scheme is

> registered with the IANA [here](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml).

Reviewer: ```suggestion
registered with the IANA [here](https://www.iana.org/assignments/uri-schemes/prov/matrix).
```

Author: Long term I'd expect that URL to change, so didn't want to have a broken link.

Currently, the `authority` and `fragment` are unused by this specification,
though are reserved for future use. Matrix does not have a central authority which
could reasonably fill the `authority` role. `nonid-segment` in the schema is
additionally reserved for future use.

The `type` denotes the kind of entity which is described by `id without sigil`.
Specifically, the following mappings are used:

* `r` for room aliases.
* `u` for users.
* `roomid` for room IDs (note the distinction from room aliases).
* `e` for events, when after a room reference (`r` or `roomid`).

{{% boxes/note %}}
During development of this URI format, types of `user`, `room`, and `event`
were used: these MUST NOT be produced any further, though implementations might

> wish to consider handling them as `u`, `r`, and `e` respectively.

Reviewer: ```suggestion
wish to consider handling them as `u`, `r`, and `e` respectively.
`roomid` will be used as indicated in development.
```

(for exhaustiveness' sake)

Author: Not sure this is important, honestly. The note isn't meant to be comprehensive, just an indicator that "weird" data might be present

    by the URI and thus is only valid on URIs which are referring to a room (it
    has no meaning and is ignored otherwise). The client should prompt for confirmation
    prior to joining the room, if the user isn't already part of the room.
  * `action=chat` - Describes an intent for the client to start/open a DM with
    the user described by the URI and thus is only valid on URIs which are referring
    to a user (it has no meaning and is ignored otherwise). Clients supporting a
    form of Canonical DMs should reuse existing DMs instead of creating new ones
    if available. The client should prompt for confirmation prior to creating the
    DM, if the user isn't being redirected to an existing canonical DM.
* `via` - Can be used to denote which servers (authorities) to attempt to resolve
  the resource through, or take `action` through. An example of using `via` for
  routing Room IDs is described [below](#routing), and is encouraged for use in
  Matrix URIs referring to a room ID. Matrix URIs can additionally use this `via`
  parameter for non-public federation resolution of identifiers (i.e.: listing a

>   server which might have information about the given user).

Author: https://github.com/matrix-org/matrix-doc/pull/3020 - github doesn't really like relative numbering in links.

This is included here and is vague because the MSC proposed it: https://github.com/matrix-org/matrix-doc/pull/2312/files#diff-f4ae31b823cdbeac105bd4b6526b3da8f6db03ae29847e1c7916b9ac51997fdfR332-R340

<!-- TODO @TR: Link to "standard identifier format" when it exists -->
Custom query parameters can be specified using the Standard Identifier Format and
appropriately encoding their values. Specifically, "percent encoding" and encoding
of the `&` are required. Where custom parameters conflict with specified ones,
clients should prefer the specified parameters. Clients should strive to maintain
consistency across custom parameters as users might be using multiple different
clients across multiple different authors. Useful and mission-aligned custom
parameters should be proposed to be included in this specification.

Examples of common URIs are:

* Room alias: `matrix:r/somewhere:example.org`
* Room ID: `matrix:roomid/somewhere:example.org?via=elsewhere.ca`

> * Permalink by room: `matrix:r/somewhere:example.org/e/event:example.org`
> * Permalink by room ID: `matrix:roomid/somewhere:example.org/e/event:example.org?via=elsewhere.ca`

Reviewer: ```suggestion
* Event permalink by room: `matrix:r/somewhere:example.org/e/abcdef123456`
* Event permalink by room ID (with routing hint): `matrix:roomid/somewhere:example.org/e/abcdef123456?via=elsewhere.ca`
```

- "Event" permalink, for explicitness' sake
- utilization of alphanumeric even IDs, as those are the new norm
- "routing hint"

Author: It's important that we keep the examples as similar as possible between the URIs. The v1 event IDs are clearer and appear less random.

Author: (though almost a year later v1 event IDs look more confusing to people - will fix)

* User: `matrix:u/alice:example.org?action=chat`

> A suggested client implementation is available in the

Reviewer 2: ```suggestion
A suggested client implementation algorithm is available in the
```
Nitpicking, I know.
